torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,195 @@
+import itertools
+from typing_extensions import Protocol
+
+import torch
+from ..parameter import UninitializedParameter
+
+
+class _LazyProtocol(Protocol):
+    """This is to avoid errors with mypy checks for 
+    The attributes in a mixin:
+    https://mypy.readthedocs.io/en/latest/more_types.html#mixin-classes
+    """
+    def _register_load_state_dict_pre_hook(self, hook):
+        ...
+
+    def register_forward_pre_hook(self, hook):
+        ...
+
+    def _lazy_load_hook(
+            self, state_dict, prefix, local_metadata, strict,
+            missing_keys, unexpected_keys, error_msgs):
+        ...
+
+    def _get_name(self):
+        ...
+
+    def _infer_parameters(self, module, input):
+        ...
+
+    @property
+    def _parameters(self):
+        ...
+
+    @property
+    def _buffers(self):
+        ...
+
+    @property
+    def _non_persistent_buffers_set(self):
+        ...
+
+    @property
+    def _load_hook(self):
+        ...
+
+    @property
+    def _initialize_hook(self):
+        ...
+
+
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Unlike regular modules, subclasses of this module can initialize
+    parameters outside of the constructor (``__init__``).
+    This allows you to, for example, initialize parameters in ``forward``
+    method to determine the shape of the weight based on the initial input.
+    Be sure to run "dummy" ``forward`` once to initialize all parameters in case
+    that you need to access the weights before the actual "real" ``forward`` pass.
+
+    >>> lazy_module = torch.nn.LazyLinear(10)
+    >>> lazy_module.forward(input)  # Dummy forward call to initialize the module parameters
+    >>> assert type(lazy_module) is torch.nn.Linear
+
+    Note that lazy modules cannot validate if the shape is correct during
```

Comment:
I appreciate what this paragraph is trying to say but this concept is complicated enough that you'd need more time and an example to explain it.

This is really a note on initialization. An example like the one I recommend above could be helpful. If both layers are lazy then the network will be initialized just like the non-lazy version. If only the first layer is lazy, however, then the initialization will be different. I think you'd need to show that here to make this point clearly.

Reply:
Added the example as you suggested :)

Reply:
Thank you!

At this point in the documentation the purpose of lazy modules as well as their basic usage pattern has been explained and demonstrated. The remainder of the doc needs to mention serialization and initialization. Serialization needs to be introduced. You could do something like this:


"Lazy modules can be serialized with a state dict like other modules. For example:

```
<brief code snippet show a lazy linear's state dict and loading from that state dict>
```

Lazy modules can also load regular :class:\`torch.nn.Parameter\`s, which replace their :class:\`torch.nn.UninitializedParameter\`s:

```
<brief code snippet showing loading initialized parameters from a lazy module after a dry run or a regular linear layer>
```

Note, however, that lazy modules cannot validate that the shape of parameters they load is correct.
"

Reply:
Ah, I need to update this, I missed this comment

Reply:
Added examples :)